fdbserver/DataDistribution.actor.cpp

```diff
@@ -893,6 +897,23 @@ struct DDTeamCollection : ReferenceCounted<DDTeamCollection> {
 		return Void();
 	}
 
+	// Returns a random healthy team, which does not contain excludeServer.
+	std::vector<UID> getRandomHealthyTeam(const UID& excludeServer) {
+		int count = 0;
+		Optional<int> idx;
+		for (int i = 0; i < teams.size(); ++i) {
+			if (teams[i]->isHealthy() && !teams[i]->hasServer(excludeServer)) {
+				if (std::rand() % ++count == 0) {
```

Comment:
The logic of this line is hard to follow. What does it do?

Also, consider using deterministicRandom() so that it is deterministic in simulation.

Reply:
Added comments.

Switched to deterministicRandom(), thanks for the tip.

Reply:
It's probably better to insert "i" into a vector as candidates and then pick a random one after this loop. The current logic doesn't guarantee to return a team, even if such a candidate exists.

Reply:
> It's probably better to insert "i" into a vector as candidates and then pick a random one after this loop. The current logic doesn't guarantee to return a team, even if such a candidate exists.

Never mind. The logic is correct and guarantee to return a team if exists.

Reply:
Actually this remind me:
If no team is healthy or there is not enough team, can we log an SevWarnAlways event?

Reply:
Changed to what jzhou77 mentioned, a random selection from the healthy team's indices, I cannot recall why I thought reservoir sampling was a gooon d idea.

Also improved the random selection to make sure if there is no healthy team, at least some team will be selected.

When no team can be found, an error will be [thrown](https://github.com/apple/foundationdb/pull/5524/files#diff-255bc4143a27b5f41ae87a1b44acdab6d2adbf7303420d9840bcd59788b9a7ebR1376), since at this point, the DD has already decided to `reboot`.